src/urllib3/response.py

```diff
@@ -909,6 +946,41 @@ def read(
 
         return data
 
+    def read1(
+        self,
+        amt: int | None = None,
+        decode_content: bool | None = None,
+    ) -> bytes:
+        self._init_decoder()
+        if decode_content is None:
+            decode_content = self.decode_content
+        if not decode_content and self._has_decoded_content:
+            raise RuntimeError(
+                "Calling read1(decode_content=False) is not supported after "
+                "read1(decode_content=True) was called."
+            )
+
+        if amt is not None and amt < len(self._decoded_buffer):
+            return self._decoded_buffer.get(amt)
```

Comment:
> Read with at most one underlying system call.  If at least one byte is buffered, return that instead.

This is from the [docstring](https://github.com/python/cpython/blob/9302f05f9af07332c414b3c19003efd1b1763cf3/Lib/http/client.py#L647-L648) of `http.client.HTTPResponse.read1`. I would follow the last sentence in urllib3 too.

So, if `self._decoded_buffer` is not empty, make no `self._raw_read` calls and:
* when `amt` is a number, return up to `amt` bytes from the buffer;
* when `amt` is `None`, return everything from the buffer.

@sethmlarson @pquentin what do you think about this?

Reply:
Avoiding calling into the kernel makes sense, I just copied the semantics of `read` which obviously don't apply here.  Have changed that condition to return early when possible.

Be interesting to see what other people think!